### PR TITLE
Improve niche placement by trying two strategies and picking the better result

### LINKED ISSUE #104807: Niche placement heuristic: place at beginning or end of type
I noticed that #102750 regressed the size of some types in the `syn` crate. I minimized the difference to the following:

```rust
pub enum Enum {
    A(A),
    B(B),
}

pub struct A {
    pub x: Thing,
    pub y: u16,
    pub z: u16,
}

pub struct B {
    pub x: Thing,
    pub y: u16,
}

pub enum Thing {
    C(u16, u16),
    D(u16, u16),
}

fn main() {
    println!("{}", std::mem::size_of::<Enum>());
}
```

```console
$ cargo +nightly-2022-11-23 run --quiet && cargo +nightly-2022-11-24 run --quiet
10
12
```

The type layout in nightly-2022-11-23 is:

```console
print-type-size type: `Enum`: 10 bytes, alignment: 2 bytes
print-type-size     variant `A`: 10 bytes
print-type-size         field `.0`: 10 bytes
print-type-size     variant `B`: 10 bytes
print-type-size         padding: 2 bytes
print-type-size         field `.0`: 8 bytes, alignment: 2 bytes
print-type-size type: `A`: 10 bytes, alignment: 2 bytes
print-type-size     field `.x`: 6 bytes
print-type-size     field `.y`: 2 bytes
print-type-size     field `.z`: 2 bytes
print-type-size type: `B`: 8 bytes, alignment: 2 bytes
print-type-size     field `.x`: 6 bytes
print-type-size     field `.y`: 2 bytes
print-type-size type: `Thing`: 6 bytes, alignment: 2 bytes
print-type-size     discriminant: 2 bytes
print-type-size     variant `C`: 4 bytes
print-type-size         field `.0`: 2 bytes
print-type-size         field `.1`: 2 bytes
print-type-size     variant `D`: 4 bytes
print-type-size         field `.0`: 2 bytes
print-type-size         field `.1`: 2 bytes
```

and in nightly-2022-11-24:

```console
print-type-size type: `Enum`: 12 bytes, alignment: 2 bytes
print-type-size     discriminant: 2 bytes
print-type-size     variant `A`: 10 bytes
print-type-size         field `.0`: 10 bytes
print-type-size     variant `B`: 8 bytes
print-type-size         field `.0`: 8 bytes
print-type-size type: `A`: 10 bytes, alignment: 2 bytes
print-type-size     field `.y`: 2 bytes
print-type-size     field `.z`: 2 bytes
print-type-size     field `.x`: 6 bytes
print-type-size type: `B`: 8 bytes, alignment: 2 bytes
print-type-size     field `.y`: 2 bytes
print-type-size     field `.x`: 6 bytes
print-type-size type: `Thing`: 6 bytes, alignment: 2 bytes
print-type-size     discriminant: 2 bytes
print-type-size     variant `C`: 4 bytes
print-type-size         field `.0`: 2 bytes
print-type-size         field `.1`: 2 bytes
print-type-size     variant `D`: 4 bytes
print-type-size         field `.0`: 2 bytes
print-type-size         field `.1`: 2 bytes
```

Graphically, the layout of Enum::A and Enum::B before and after are:

<table>
<tr><th></th><th>Before</th><th>After</th></tr>
<tr><th>Enum::A</th><td><pre>
+-----+-----+-----+-----+-----+
| dsc | u16 | u16 | u16 | u16 |
+-----+-----+-----+-----+-----+
^~~Thing~~~~~~~~~~^
^~~A~~~~~~~~~~~~~~~~~~~~~~~~~~^
</pre></td><td><pre>
+-----+-----+-----+-----+-----+-----+
| dsc | u16 | u16 | dsc | u16 | u16 |
+-----+-----+-----+-----+-----+-----+
                  ^~~Thing~~~~~~~~~~^
      ^~~A~~~~~~~~~~~~~~~~~~~~~~~~~~^
</pre></td></tr>
<tr><th>Enum::B</th><td><pre>
      +-----+-----+-----+-----+
..... | dsc | u16 | u16 | u16 |
      +-----+-----+-----+-----+
      ^~~Thing~~~~~~~~~~^      
      ^~~B~~~~~~~~~~~~~~~~~~~~^
</pre></td><td><pre>
+-----+-----+-----+-----+-----+
| dsc | u16 | dsc | u16 | u16 | .....
+-----+-----+-----+-----+-----+
            ^~~Thing~~~~~~~~~~^
      ^~~B~~~~~~~~~~~~~~~~~~~~^
</pre></td></tr>
</table>

Notice how the old layout is putting `Thing` at the beginning of `A` and `B`, while the new layout is putting `Thing` at the end of `A` and `B`. The reason the new layout is worse is that now, when building `Enum`, there is no way to line up `A`'s existing niche with the `B` case's padding. Instead a whole new discriminant needs to be added.

From reading the description of #102750, I don't get the impression that the justification of that PR applies to the difference in this placement of `Thing` inside `A` and `B` in the code above. It seems like just a coincidence of the implementation that the PR affected this code.

My observation is that, other things being equal, we should prefer to locate niches ~~in the very beginning or very end of a type, not in the middle~~ near the beginning or end of a type, as far from the middle of the type as possible. This will allow them to get lined up with the padding of smaller enum variants. Placing a niche as far from the middle of the type as possible makes room for the largest possible other type to line up before or after the niche.

Mentioning @the8472 @wesleywiser since you were recently working on layout.

### LINKED ISSUE #105371: #102750 regresses layout size of Cow<>
```rust
    assert_eq!(std::mem::size_of::<std::borrow::Cow<'static, str>>(), 24);
```

Passes on stable and beta but not on Nightly.

I bisected it, results below which point to #102750:

searched nightlies: from nightly-2022-10-20 to nightly-2022-12-06
regressed nightly: nightly-2022-11-24
searched commit range: https://github.com/rust-lang/rust/compare/ff8c8dfbe66701531e3e5e335c28c544d0fbc945...70f8737b2f5d3bf7d6b784fad00b663b7ff9feda
regressed commit: https://github.com/rust-lang/rust/commit/4e0d0d757e2f1b61ec809420b006545a9f8974c0

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.5</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start 2022-10-20 
```
</details>

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

cc: @nnethercote @the8472 @wesleywiser

### PR DESCRIPTION
**Pull request by the8472**
Fixes #104807
Fixes #105371

Determining which sort order is better requires calculating the struct size (so we can calculate the niche offset). But that in turn depends on the field order, so happens after sorting. So the simple way to solve that is to run the whole thing twice and pick the better result.

1st commit is just code motion, the meat is in the later ones.

### CODE REVIEW

File: tests/codegen/issue-103840.rs [hunk 1]
```diff
@@ -1,4 +1,8 @@
+// should-fail
+// known-bug: #103840
 // compile-flags: -O
+// FIXME: reenable after next llvm update. currently the codegen is too fickle, minor layout
+// perturbations in Vec break this test.
```
Comment (the8472) [inline tests/codegen/issue-103840.rs:5]: see also #106790 and nikic's comment there https://github.com/rust-lang/rust/pull/106790#issuecomment-1382904714

File: tests/ui/async-await/future-sizes/async-awaiting-fut.stdout [hunk 2]
```diff
@@ -2,38 +2,34 @@ print-type-size type: `[async fn body@$DIR/async-awaiting-fut.rs:21:21: 24:2]`:
...
 print-type-size     variant `Unresumed`: 0 bytes
 print-type-size     variant `Suspend0`: 3077 bytes
-print-type-size         local `.__awaitee`: 3077 bytes, offset: 0 bytes, alignment: 1 bytes
+print-type-size         local `.__awaitee`: 3077 bytes
 print-type-size     variant `Returned`: 0 bytes
 print-type-size     variant `Panicked`: 0 bytes
 print-type-size type: `[async fn body@$DIR/async-awaiting-fut.rs:10:64: 19:2]`: 3077 bytes, alignment: 1 bytes
 print-type-size     discriminant: 1 bytes
-print-type-size     variant `Unresumed`: 2051 bytes
-print-type-size         padding: 1026 bytes
```
[the remark this answers is not in the document]
Comment (the8472) [reply tests/ui/async-await/future-sizes/async-awaiting-fut.stdout:11]: Hrm, not sure how to judge that one. It's a variant in an explicit-discriminant enum and I didn't pay much attention to those since they don't benefit from niche optimization. At least in this particular case it doesn't affect the overall enum size.

I also don't understand the meaning of `, offset: 0 bytes, alignment: 1 bytes` disappearing on other variants.

Should I investigate this?

[the remark this answers is not in the document]
Comment (Nilstrieb) [reply tests/ui/async-await/future-sizes/async-awaiting-fut.stdout:11]: This is a big improvement for the variant, so nothing to investigate. Although yeah, it doesn't matter at all (and could even just be print-type-size variant size calculation being odd).

File: compiler/rustc_abi/src/layout.rs [hunk 3]
```diff
@@ -934,3 +775,287 @@ pub trait LayoutCalculator {
...
     }
 }
+
+fn univariant(
+    this: &(impl LayoutCalculator + ?Sized),
+    dl: &TargetDataLayout,
+    fields: &IndexSlice<FieldIdx, Layout<'_>>,
+    repr: &ReprOptions,
+    kind: StructKind,
+    niche_bias_start: bool,
```
Comment (WaffleLapkin) [inline compiler/rustc_abi/src/layout.rs:785]: I think the code could benefit from this being an `enum` with documentation explaining exactly what it decides

File: compiler/rustc_abi/src/layout.rs [hunk 4]
```diff
@@ -934,3 +775,287 @@ pub trait LayoutCalculator {
...
+                            })
+                        };
+
+                        // Place ZSTs first to avoid "interesting offsets", especially with only one
+                        // or two non-ZST fields. This helps Scalar/ScalarPair layouts.
+                        // Then place largest alignments first.
+                        // Then prioritize niche placement within alignment group according to
+                        // `niche_bias_start`.
+                        // Then among fields with equally-sized niches prefer the ones
+                        // closer to the start/end of the field.
```
Comment (WaffleLapkin) [inline compiler/rustc_abi/src/layout.rs:859]: Why does this matter?

File: compiler/rustc_abi/src/layout.rs [hunk 5]
```diff
@@ -934,3 +775,287 @@ pub trait LayoutCalculator {
...
+                    //
+                    // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
+                    let align = layout.align().abi.bytes();
+                    let size = layout.size().bytes();
+                    let size_as_align = align.max(size).trailing_zeros();
+                    let size_as_align = if any_niche {
+                        max_field_align.trailing_zeros().min(size_as_align)
+                    } else {
+                        size_as_align
+                    };
```
Comment (WaffleLapkin) [inline compiler/rustc_abi/src/layout.rs:829]: Why is this needed? Tbh I'm confused by this whole closure...

Also it looks like you don't need to compute `max_field_align` if `!any_niche`, might be a good idea to do something like `let max_field_align = any_niche.then(|| ...);`

File: compiler/rustc_abi/src/layout.rs (hunk 4 above)
Comment (the8472) [reply compiler/rustc_abi/src/layout.rs:859]: Let's say we have

```
#[repr(C)]
struct A(u8, bool, u8);

struct B(bool);

struct C(A, B);
```

Then we have two niche-candidates (both bools). Since we try to get the niches close to the edge of a struct we try to prioritize B in this case because its niche sits closer to the start/end of the field.

File: compiler/rustc_abi/src/layout.rs (hunk 5 above)
Comment (the8472) [reply compiler/rustc_abi/src/layout.rs:829]: Given

```
struct A(u8, [u8; 16]);

struct B(bool, [u8; 16]);
```

We want A to move the array to the front so it has a chance to be aligned when embedded in a struct with higher alignment. This allows aligned loads.

But for B we want to prioritize the bool to get a niche at offset 0.

File: compiler/rustc_abi/src/layout.rs (hunk 4 above)
Comment (WaffleLapkin) [reply compiler/rustc_abi/src/layout.rs:859]: But why do we want niches close to the start/end?

Comment (the8472) [reply compiler/rustc_abi/src/layout.rs:859]: dtolnay has a good explanation in #104807

If the niche is at the edge then it's more likely that the niche-as-discriminant encoding will end up being beneficial. Otherwise it'll pick the explicit-discriminant layout which results in larger enums. Or at least it'll have to arrange the variants in a less compact way.

File: compiler/rustc_abi/src/layout.rs (hunk 5 above)
Comment (the8472) [reply compiler/rustc_abi/src/layout.rs:829]: Start is generally preferred because it doesn't suffer from alignment padding.
End is 2nd choice and useful when it's in the largest variant in an enum.

Comment (the8472) [reply compiler/rustc_abi/src/layout.rs:829]: > Also it looks like you don't need to compute max_field_align if !any_niche, might be a good idea to do something like let max_field_align = any_niche.then(|| ...);

We make many passes over the array anyway, including sorting, so probably not worth optimizing it. And I think it would hurt readability or require additional explanation if it only gets calculated conditionally.

**Comment by the8472** [comment]
- turned the flag into an enum
- added more comments
- refined the logic of `alignment_group_key`

**Review by wesleywiser — APPROVED** [review]
This looks good to me. Thanks for adding the additional comments in they layout algorithm and test cases! 

r=me with final green perf run

**Comment by the8472** [comment]
Results look fairly mixed but there are a few big wins on instructions, hope that counts as sufficiently greenish. And it does fix the type size regressions that were reported by users.

@bors r=wesleywiser

**Comment by the8472** [comment]
fixed the test depending on word-size.

@bors r=wesleywiser

**Comment by the8472** [comment]
excluded 32bit targets on a codegen tests because llvm seems to make different choices there (memset vs. stores)

@bors r=wesleywiser

**Comment by pnkfelix** [comment]
* addresses a real performance regression (#105371) to how we laid out niches.
* primary big benefit to syn. rest of primary changes may be noise; the secondary improvements are nice to have.
* marking as triaged.

@rustbot label: +perf-regression-triaged
